source/auth/auth.rst

```diff
@@ -453,10 +573,77 @@ Implementation
 Test Plan
 =========
 
-Tests have been defined in the associated files:
+Connection string tests have been defined in the associated files:
 
 * `Connection String <tests/connection-string.json>`_.
 
+---------------------------------------
+SCRAM-SHA-256 and mechanism negotiation
+---------------------------------------
+
+Testing SCRAM-SHA-256 requires server version 3.7.3 or later with
+``featureCompatibilityVersion`` of "4.0" or later.
+
+Drivers that allow specifying auth parameters in code as well as via
+connection string should test both for the test cases described below.
+
+Step 1
+------
+
+Create three test users, one with only SHA-1, one with only SHA-256 and one
+with both.  For example::
+
+    db.createUser({user: 'sha1', pwd: 'sha1', roles: ['root'], mechanisms: ['SCRAM-SHA-1']})
+    db.createUser({user: 'sha256', pwd: 'sha256', roles: ['root'], mechanisms: ['SCRAM-SHA-256']})
+    db.createUser({user: 'both', pwd: 'both', roles: ['root'], mechanisms: ['SCRAM-SHA-1', 'SCRAM-SHA-256']})
+
+Optional, but strongly encouraged: Create a SCRAM-SHA-256 user with a username
+and password that are modified by SASLprep, for example ``"\u2168"`` (ROMAN
+NUMERAL NINE), which is transformed by SASLprep to ``IX``.  Note: when calling
+the ``createUser`` command, the username must be prepped or the command will
+fail.  The password need not be prepped.
```

Comment:
Is it still true that the username must be prepped, or did Bernie request a server behavior change?

Reply:
It is still true today. We had a meeting to formulate an alternative approach this morning, which will require considerable changes on the server. For now, saslprep as documented here.

Reply:
For the purpose of testing SCRAM-SHA-256 (but not negotiation where both are possible), creating with prepped is sufficient.

Reply:
Let's drop this for now, it's confusing the heck out of me. The test is encouraged but not required, and the password could be prepped but "need not be prepped"?

Let's save this for a separate patch once the server behavior is settled. Once we know how to create a user with a non-ASCII username and password then we'll definitely need to test that we can authenticate with those non-ASCII credentials, otherwise we won't know if we implemented SASLPrep at all. We'll need to specify an update drivers' createUser helper methods at that time, too.

Reply:
I have added a new test section specific to SASLprep that can be done with today's implementation.